### PR TITLE
pytorch: renamed from libtorch

### PR DESCRIPTION
**Pull request by singingwolfboy**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----
This renames `libtorch` to `pytorch`, and installs the Python module as well as the C++ libraries. This was originally #110199, but I opened a new pull request to get more attention on this change.

### CODE REVIEW

File: Formula/pytorch.rb
Comment (carlocab) [inline Formula/pytorch.rb:28]: I'm fine with the rename, but I'm not fine with making this depend on ARM.

If it fails to build on Intel, then we need to understand why and fix it. Or, at least, report the problem upstream so that it can be fixed there.

Comment (singingwolfboy) [reply Formula/pytorch.rb:28]: I just reported the problem upstream: https://github.com/pytorch/pytorch/issues/85956

Comment (carlocab) [reply Formula/pytorch.rb:28]: Thanks!

**Comment by danielnachun** [comment]
@singingwolfboy when https://github.com/Homebrew/homebrew-core/pull/110116 is merged could you rebase this?  It should now work on Intel macOS so you won't need to have the ARM requirement any more.  That was the only blocker we seemed to have over this.

I added some of your changes in that PR as well so you shouldn't have to do much other than the rename itself and tweaks you made to the build parameters and test.

**Comment by singingwolfboy** [comment]
Ugh, still failing! 😩 This is the new error message, still only on Intel Macs:

```
error: unknown warning option '-Wno-unused-but-set-variable'; did you mean '-Wno-unused-const-variable'? [-Werror,-Wunknown-warning-option]
```

Does anyone know what might be causing this, or how to fix it?

**Comment by danielnachun** [comment]
> Ugh, still failing! 😩 This is the new error message, still only on Intel Macs:
> 
> ```
> error: unknown warning option '-Wno-unused-but-set-variable'; did you mean '-Wno-unused-const-variable'? [-Werror,-Wunknown-warning-option]
> ```
> 
> Does anyone know what might be causing this, or how to fix it?

That's odd because we literally just built this with no issues and https://github.com/Homebrew/homebrew-core/pull/112520 seems to have gotten past building `fbgemm` at least, because it timed out after 90 minutes.  It might be worth merging that PR first and rebasing here again to minimize the changes you are making to the `libtorch` formula aside from the renaming itself.  That can help us narrow down why that flag is erroneously being used when it doesn't exist.

File: Formula/pytorch.rb
Comment (danielnachun) [inline Formula/pytorch.rb:47]: I think instead we need to backport the part of https://github.com/pytorch/pytorch/commit/d7e6aaa59b4660876d3d9bbee50113fad2c0b115 that changes 
```
target_compile_options(asmjit -Wno-unused-but-set-variable)
```
to
```
target_compile_options_if_supported(asmjit -Wno-unused-but-set-variable)
```
That will make sure that flag is only added if the compile supports the flag, where as in the current release it uses the flag unconditionally.

Comment (singingwolfboy) [reply Formula/pytorch.rb]: I thought about doing that, but unfortunately the `target_compile_options_if_supported` macro is not present in the 1.12.1 release! It's a custom macro that was committed after the release was created. So we would have to patch that function back in as well, which seems like a bigger change than necessary.

I think maybe we can just delete these lines entirely? That's what I'm trying now in my latest commit.

Comment (singingwolfboy) [reply Formula/pytorch.rb:47]: Nope, patches aren't allowed in new formulae, and brew-test-bot sees this as a new formula (even though it's a rename, which is different). So I'm going to try patching it to `if(FALSE)` instead for the same effect.

Comment (danielnachun) [reply Formula/pytorch.rb:47]: It looks like this worked now and just timed out, though we should only do the `inreplace` if the macOS version is less than or equal to Catalina.

Comment (singingwolfboy) [reply Formula/pytorch.rb:47]: Why less than or equal to Catalina? Also, could you please add the "long build" label to this pull request, so that it doesn't time out for the next CI build?

Comment (danielnachun) [reply Formula/pytorch.rb:47]: Actually let's do a CI run to see if all platforms can build with the `if(FALSE)` fix used.  We'll be able to drop it in the next release anyway.

Comment (danielnachun) [reply Formula/pytorch.rb:47]: CI worked at last!  Great job hunting down that fix, and thanks for your patience.  This PR was beset by multiple unrelated problems that were non-trivial to fix but I think we're ready to go.  I will merge this tomorrow if there has been no other maintainer feedback as the other changes here are pretty straight forward.

File: Formula/pytorch.rb
Comment (danielnachun) [inline Formula/pytorch.rb:81]: I don't think it should block this PR, but I think we should report upstream that there should be a way to set the paths to `torch_shm_manager` and `_TORCH_PATH` as CMake options.  Hardcoding them like this reduces the portability of this package and these types of paths are usually pretty easy to pass in as options.

**Review by fxcoudert — APPROVED** [review]

**Comment by fxcoudert** [comment]
Thank you @singingwolfboy for this pull request!

**Pull request closed without merging**
